### PR TITLE
:sparkles: Adding caching and scoping to the analysis rpc server

### PR DESCRIPTION
**Pull request by shawn-hurley**
* Fixing similar issues getting kicked out of priority 0 stack
* Fixed and issue where similar tasks not found for analyzer tasks

### CODE REVIEW

File: kai_analyzer_rpc/go.mod [hunk 1]
```diff
@@ -5,7 +5,8 @@ go 1.23.1
 require (
 	github.com/go-logr/logr v1.4.2
 	github.com/konveyor/analyzer-lsp v0.5.3
-	github.com/shawn-hurley/java-external-provider v0.5.4-0.20240904135407-b0958af4ebdf
+	github.com/konveyor/analyzer-lsp/external-providers/java-external-provider v0.0.0-20241107203043-172d684c168d
```
Comment (shawn-hurley) [inline kai_analyzer_rpc/go.mod:8]: Need to update this once the other PR's merge

File: kai_analyzer_rpc/main.go [hunk 2]
```diff
@@ -23,7 +23,8 @@ func main() {
 	depOpenSourceLabelsFile := flag.String("depOpenSourceLabelsFile", "", "Path to the dep open source labels file")
 
 	// TODO(djzager): We should do verbosity type argument(s)
-	logLevel := slog.LevelInfo
+	//logLevel := slog.LevelInfo
+	logLevel := -6
```
Comment (shawn-hurley) [inline kai_analyzer_rpc/main.go:27]: Probably need to turn this back, sorry

File: kai/reactive_codeplanner/agent/analyzer_fix/agent.py [hunk 3]
```diff
@@ -131,7 +131,7 @@ def execute(self, ask: AgentRequest) -> AnalyzerFixResponse:
         resp = self.parse_llm_response(aimessage)
         return AnalyzerFixResponse(
             encountered_errors=[],
-            file_to_modify=ask.file_path,
+            file_to_modify=ask.file_path.absolute(),
```
Comment (JonahSussman) [inline kai/reactive_codeplanner/agent/analyzer_fix/agent.py:134]: I think this should be `.resolve()` not `.absolute()`. `.absolute()` doesn't transform `..` parts into references to the parent directory or resolve symbolic links:

```python
from pathlib import Path
import os

print(os.getcwd())
# /home

p = Path("../thing")

print(p.absolute())
# /home/../thing
print(p.resolve())
# /thing
```

File: kai/reactive_codeplanner/task_runner/analyzer_lsp/task_runner.py [hunk 4]
```diff
@@ -50,8 +50,9 @@ def execute_task(self, rcm: RepoContextManager, task: Task) -> TaskResult:
         with open(task.file) as f:
             src_file_contents = f.read()
 
+        logger.info(f"file -- {task.file}")
         agent_request = AnalyzerFixRequest(
-            file_path=Path(task.file),
+            file_path=Path(task.file).absolute(),
```
Comment (JonahSussman) [inline kai/reactive_codeplanner/task_runner/analyzer_lsp/task_runner.py:55]: Same comment as above.

File: kai/reactive_codeplanner/task_runner/analyzer_lsp/validator.py [hunk 5]
```diff
@@ -80,10 +81,10 @@ def __init__(self, config: RpcClientConfig) -> None:
         super().__init__(config)
 
     @tracer.start_as_current_span("analyzer_run_validation")
-    def run(self) -> ValidationResult:
+    def run(self, scoped_paths: Optional[list[Path]]) -> ValidationResult:
```
Comment (JonahSussman) [inline kai/reactive_codeplanner/task_runner/analyzer_lsp/validator.py:84]: Might want to default `run` and `__run_analyzer_lsp`'s optional arguments to `None` if they're truly optional.

File: kai/reactive_codeplanner/task_runner/compiler/maven_validator.py [hunk 6]
```diff
@@ -28,7 +28,7 @@ def __init__(self, config: RpcClientConfig) -> None:
         self.last_compilation_errors: list[MavenCompilerError] = []
 
     @tracer.start_as_current_span("maven_run_validator")
-    def run(self) -> ValidationResult:
+    def run(self, scoped_paths: Optional[list[Path]]) -> ValidationResult:
```
Comment (JonahSussman) [inline kai/reactive_codeplanner/task_runner/compiler/maven_validator.py:31]: Same as above

File: tests/test_task_manager.py [hunk 7]
```diff
@@ -20,7 +21,7 @@ def __init__(
         self.error_sequences = error_sequences
         self.run_count = 0
 
-    def run(self) -> ValidationResult:
+    def run(self, scoped_paths: Optional[list[str]]) -> ValidationResult:
```
Comment (JonahSussman) [inline tests/test_task_manager.py:24]: Same as above

File: example/run_demo.py [hunk 8]
```diff
@@ -216,6 +215,7 @@ def run_demo(report: Report, server: JsonRpcServer) -> None:
         for count, (file_path, incidents) in enumerate(impacted_files.items(), 1):
             for incident in incidents:
                 incident.uri = os.path.join(SAMPLE_APP_DIR, file_path)
+                incident.uri = Path(incident.uri).absolute().__str__()
```
Comment (JonahSussman) [inline example/run_demo.py:218]: See below for comment about `.absolute()`.

**Review by JonahSussman — COMMENTED** [review]
Looks good, a few things here and there

File: example/run_demo.py [hunk 9]
```diff
@@ -106,10 +106,7 @@ def initialize_rpc_server(
             log=log,
         ),
         app=app,
-        # TODO(fabianvf): when bumping the iterations/depth/priority, it can increase
-        # execution time significantly. We need to add some kind of keepalive signal or
-        # something to prevent a low timeout from killing a properly working request.
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply example/run_demo.py:111]: I don't think so, but my solution is, we watch the logs for right now, especially in run_demo, the logs are in the same window. 

For the IDE integration, I assume we need $SOMETHING, and I thought that was being tracked in the notification/intra-run update

File: kai/reactive_codeplanner/agent/analyzer_fix/agent.py (hunk 3 above)
Comment (shawn-hurley) [reply kai/reactive_codeplanner/agent/analyzer_fix/agent.py:134]: do we really want to resolve symlinks though?

File: Makefile
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply Makefile:1]: updated

File: Makefile [hunk 10]
```diff
@@ -48,8 +50,6 @@ get_analyzer_deps:
 get_rulesets:
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply Makefile:56]: Updated

File: kai/reactive_codeplanner/agent/analyzer_fix/agent.py (hunk 3 above)
Comment (shawn-hurley) [reply kai/reactive_codeplanner/agent/analyzer_fix/agent.py]: updated to use `os.path.abspath` to expand the relative path things like `..` but not resolve symlinks.

File: kai/reactive_codeplanner/task_runner/analyzer_lsp/validator.py (hunk 5 above)
Comment (shawn-hurley) [reply kai/reactive_codeplanner/task_runner/analyzer_lsp/validator.py]: updated

File: tests/test_task_manager.py (hunk 7 above)
Comment (shawn-hurley) [reply tests/test_task_manager.py]: updated

File: example/run_demo.py (hunk 8 above)
Comment (shawn-hurley) [reply example/run_demo.py]: updated, see above comment.

File: example/run_demo.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JonahSussman) [reply example/run_demo.py:111]: Gotcha gotcha. The intra-run update still needs work AFAIK, but I'm good with removing this comment for now.

**Review by JonahSussman — APPROVED** [review]
Not a Go expert, but LGTM

**Review by pranavgaikwad — APPROVED** [review]

**Pull request merged**